src/libasr/casting_utils.cpp

```diff
@@ -146,6 +146,14 @@ namespace LCompilers::CastingUtil {
         } else {
             dest = ASRUtils::extract_type(dest);
         }
+        if (ASR::is_a<ASR::ArrayConstant_t>(*expr)) {
```

Comment:
I think casting should be done only on the basis of `type`. Using expression to decide casting operation seems more like a hard-coded approach.